### DISCUSSION TITLE
Reuse the ZAP (ZCL Advanced Platform) generic templating/generation engine?

### DISCUSSION THREAD

**Discussion by Hedda** [ffenix113/zigbee_home#4] [Ideas]
Probably overkill but have you considered looking at repurposing ZAP (ZCL Advanced Platform) for UI and generate device templates?

https://github.com/project-chip/zap

It is based on Zigbee Cluster Library; "_It is used to configure clusters, attributes and other entities for Matter and Zigbee Pro applications_"

https://docs.silabs.com/simplicity-studio-5-users-guide/latest/ss-5-users-guide-developing-with-project-configurator/zigbee-cluster-configurator

https://docs.silabs.com/zap-tool/1.0.0/zap-users-guide/

https://github.com/project-chip/zap/blob/master/README.md

# ZCL Advanced Platform

[![Build and release packages](https://github.com/project-chip/zap/actions/workflows/release.yml/badge.svg)](https://github.com/project-chip/zap/actions/workflows/release.yml)
[![Zigbee code regeneration](https://github.com/project-chip/zap/actions/workflows/zigbee.yml/badge.svg)](https://github.com/project-chip/zap/actions/workflows/zigbee.yml)
[![Matter code regeneration](https://github.com/project-chip/zap/actions/workflows/matter.yml/badge.svg)](https://github.com/project-chip/zap/actions/workflows/matter.yml)

## [What is ZAP?](https://docs.silabs.com/zap-tool/1.0.0/zap-start/)

ZAP is a generic generation engine and user interface for applications and libraries based on Zigbee Cluster Library, the specification developed by the [Connectivity Standards Alliance](https://csa-iot.org/).

ZAP allows you to perform the following:

- perform SDK-specific customized generation of all global artifacts (constants, types, IDs, etc) based on the ZCL specification
- provide UI for the end-user to select specific application configuration (clusters, attributes, commands, etc.)
- perform SDK-specific customized generation of all user selected configuration artifacts (application configuration, endpoint configuration, etc) based on ZCL specification and customer-provided application configuration.

ZAP is a generic templating engine. Examples are provided for how to generate artifacts for the C language environment, but one could easily add new templates for other language environments, such as C++, java, node.js, python or any other.

## Quick setup

Refer to [Setup Instructions](https://docs.silabs.com/zap-tool/1.0.0/zap-getting-started/zap-installation)

### Source code

Refer to [development instructions](docs/development-instructions.md) for more details.

## License

This software is licensed under [Apache 2.0 license](LICENSE.txt).

## Usage Documentation

- [ZAP Fundamentals](https://docs.silabs.com/zap-tool/1.0.0/zap-fundamentals/)
- [ZAP User's Guide](https://docs.silabs.com/zap-tool/1.0.0/zap-users-guide/)

## Detailed Developer Documentation

- [FAQ](docs/faq.md)
- [Release instructions](docs/release.md)
- [Development Instructions](docs/development-instructions.md)
- [Design](docs/design.md)
- [Template tutorial](docs/template-tutorial.md)
- [SDK integration guideline](docs/sdk-integration.md)
- [Custom ZCL entities design](docs/custom-zcl.md)
- [API](docs/api.md)
- [Coding standard](docs/coding-standard.md)

**Comment by Hedda** [comment]
The backstory is that the project-chip originally developed ZAP (ZCL Advanced Platform) generic templating/generation engine to make DIY devices since there initially were no commercial Thread based Matter devices available that you can buy as of yet (though some manufacturers said that they might offer optional firmware to migrate newer Zigbee devices to Thread and Matter). Anyway, thus developers  use ZAP to make their own custom Thread based Matter devices using Nordic Semiconductor's nRF52840 Dongle and their nRF5 SDK:

https://docs.nordicsemi.com/search?q=zap

**Comment by ffenix113** [comment]
Sounds like a useful project! As this project becomes more stable I am sure that this can be beneficial to make configuration easier, or even allow support for Matter.

Thank you for the links and the idea.
